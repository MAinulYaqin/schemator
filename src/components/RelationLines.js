import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { connect } from "react-redux";

import { getPathPoints } from "../helpers/layout";

const RelationLine = styled.path`
  fill: none;
  stroke: white;
  stroke-width: 2px;
`;

export const RelationLines = ({ tables, fields, relations }) => {
  const byTableID = tableID => item => item.tableID === tableID;
  const byID = itemID => item => item.id === itemID;

  return (
    <g>
      {relations.map(relation => {
        const { fieldID, fromTableID, toTableID } = relation;
        const fieldIndex = fields
          .filter(byTableID(fromTableID))
          .findIndex(byID(fieldID));
        const fromTable = tables.find(byID(fromTableID));
        const toTable = tables.find(byID(toTableID));

        if (fromTable && toTable) {
          const { position: fPos } = fromTable;
          const { position: tPos } = toTable;
          const points = getPathPoints(fPos, tPos, fieldIndex);

          return <RelationLine key={relation.id} d={points} />;
        }

        return null;
      })}
    </g>
  );
};

RelationLines.propTypes = {
  tables: PropTypes.array,
  fields: PropTypes.array,
  relations: PropTypes.array
};

export const mapStateToProps = ({ tables, fields, relations }) => ({
  tables,
  fields,
  relations
});

export default connect(mapStateToProps)(RelationLines);
