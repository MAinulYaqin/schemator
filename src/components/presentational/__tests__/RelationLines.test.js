import React from "react";
import { shallow } from "enzyme";

import RelationLines from "../RelationLines";

describe("RelationLines", () => {
  const setup = propOverrides => {
    const props = {
      items: [
        {
          x1: 32,
          y1: 32,
          x2: 128,
          y2: 128
        },
        {
          x1: 240,
          y1: 32,
          x2: 160,
          y2: 128
        }
      ],
      ...propOverrides
    };

    const wrapper = shallow(<RelationLines {...props} />);

    return {
      wrapper,
      props
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });
});
