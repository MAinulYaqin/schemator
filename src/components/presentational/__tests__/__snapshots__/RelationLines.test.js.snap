// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelationLines should renders default 1`] = `
<g>
  <styled.path
    d="
          M566 877 
          C484 877 
            396 128 
            332 128
        "
    key="5dde1193-9c8f-4c45-bb1e-58455cb2a7d8"
  />
</g>
`;
